pippy/PipelineDriver.py

```diff
@@ -593,32 +592,63 @@ def extract_value_ref_args(arg):
             for fut in _futures:
                 fut.wait()
 
-        with self.value_store_lock:
+        with self.value_store_cv:
             assert output_unique_key not in self.value_store
             self.value_store[output_unique_key] = RefcountedFuture(future, output_refcount)
+            self.value_store_cv.notify_all()
 
-        return ValueReference(self.stage_id, output_unique_key)
+        finish_ts = time.time()
+        self.record_event(rank=self.rank_worker.rank, start_ts=start_ts, finish_ts=finish_ts, id=id, name=name, type='received', mbid=cur_microbatch)
+        self.record_event_dependency(from_id=name, to_id=target_name, type='waiting')
 
-    def index_value(self, output_unique_key : str, output_refcount : int, value_ref, idx):
-        # For the purposes of refcounting, decrement this use
-        with self.value_store_lock:
-            refcounted_future = self.value_store[value_ref.unique_key]
-            if refcounted_future.release():
-                self.value_store.pop(value_ref.unique_key)
+        return ValueReference(self.stage_id, output_unique_key)
 
-            indexed = refcounted_future.future.then(lambda f: f.value()[idx])
+    def coalesce_index_value(self, indices : List[Tuple]):
+        for index_tuple in indices:
+            (output_unique_key, output_refcount, value_ref, idx) = index_tuple
+            logging.info(f'[{self.stage_id}] Received getitem call: {(output_unique_key, output_refcount, value_ref, idx)}')
+            with self.value_store_cv:
+                if output_unique_key in self.value_store:
+                    # TODO: investigate why there are repeated getitem calls with 0 users
```

Comment:
Make an issue

Also this looks to be guarded by an assert. Are there situations the user would run into where this assert is false and this tears down the program? would that be a user-facing bug/failure?

Reply:
First of all, I guess there should not be repeated index calls. Should there be?
Second, the repeated calls I have seen (with T5 only) all have 0 users, so I am just letting it pass **for now** as no other stage is using it.
But, if there is indeed a repeated call with non-zero users, then I would like it to error out and let user report to us.
Thus, I added an assertion error message in a new commit:
"PiPPy internal error: repeated index value call with non-zero users"